enhancements/network/static-ip-addresses-vsphere.md

```diff
@@ -0,0 +1,568 @@
+
+This enhancement expands on the [`hosts` slice](https://github.com/openshift/enhancements/blob/master/enhancements/network/baremetal-ipi-network-configuration.md#user-facing-api) by associating hosts with roles and, optionally, a failure domain.
+
+For the bootstrap and control plane nodes, static IPs are passed to the node via the `guestinfo.afterburn.initrd.network-kargs` extraconfig parameter.  [Afterburn](https://github.com/coreos/afterburn/blob/main/src/providers/vmware/amd64.rs) recognizes this parameter when the node initially boots. 
+
+When static IP configuration is required, Machines can not be created via MachineSets.
+The installer must create the initial set of compute Machines manually and an administrator
+must implement a `preCreate` hook for the MachineSet to allow the MachineSet to create
+Machines during day-2 operations. See [IP Configuration of Machines](#Scaling-new-Nodes-with-`machinesets`) for more information.
+
+As with the installer, the vSphere [machine reconciler](https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/vsphere/reconciler.go#L745-L755) 
+will pass the static IP configuration via the `guestinfo.afterburn.initrd.network-kargs` extraconfig parameter.  
+
+
+### Day 2 Static IP Configuration
+
+Nodes being added to a cluster may be configured via an `nmstate` IP configuration or default to DHCP.  The networking configuration of a node/machine is immutable after creation. The vSphere machine API machine controller will apply the nmstate configuration when the associated VM is cloned.
+
+`machinesets` will be supported through the creation of a user-created custom controller.  This custom controller will leverage machine lifecycle hooks to
+provide IP configuration to machines descending from `machinesets` with `machine` annotated with the appropriate lifecycle hook.
+
+#### Changes Required
+
+##### Installer
+
+1. Modify the `install-config.yaml` vSphere platform specification to support nmstate configuration. The nmstate configuration is intended to be structurally compatible with the [nmstate API](https://nmstate.io/) but will omit fields that do not map logically to installation of a cluster.
+
+~~~go
+// Hosts defines `Host` configurations to be applied to nodes deployed by the installer
+type Hosts struct {
+  // Bootstrap IP configuration for the bootstrap node. If omitted, node will default to DHCP.
+  // +optional
+  Bootstrap *Host `json: "bootstrap,omitempty"`
+
+  // ControlPlane slice of IP configurations for the control plane nodes. If empty, nodes will default to DHCP.
+  // +optional
+  ControlPlane []Host `json: "controlPlane,omitempty"`
+
+  // Compute slice of IP configurations for the compute nodes. If empty, nodes will default to DHCP.
+  // +optional
+  Compute []Host `json: "compute,omitempty"`
+}
```

Comment:
I went back and forth on where I thought this config should live as well.  I feel on putting it in the platform spec as it aligned with bare metal and allowed the bootstrap to be specifically defined.  I do think putting it in the machine pool does make a lot of sense too though.  I'm certainly fine with making that adjustment.

Reply:
Upon thinking about this further last night and this morning, keeping this in the platform spec would allow failure domains to define their own static IP configurations, which would be needed for supporting static IPs in multi-zone.

Reply:
> Upon thinking about this further last night and this morning, keeping this in the platform spec would allow failure domains to define their own static IP configurations, which would be needed for supporting static IPs in multi-zone.

Actually, my apologies on that, we do have a failure domain field in the host definition.  We can just move this under the machine pools.  Will make that change.

Reply:
The most consistent with baremetal I think would be to have a single list of hosts with a 'role' field in `installconfig.platform.vsphere.hosts`, rather than separate bootstrap/controlplane/compute lists.
Putting it in the MachinePool feels slightly weird to me, because that is so explicitly about configuring the pool of hosts collectively, not individually. But maybe that's just historical bias on my part.

Reply:
I do like the idea of introducing a `role` field rather than having separate slices for each.  @patrickdillon based on @zaneb feedback, do you feel strongly about this living in the machine pools?  I do think maintaining consistency with baremetal would be a pro for putting this in the platform spec.